R/scale-colour.r

```diff
@@ -104,29 +108,68 @@ scale_fill_continuous <- function(...,
   if (is.function(type)) {
-    type(...)
-  } else if (identical(type, "gradient")) {
-    scale_colour_steps(...)
-  } else if (identical(type, "viridis")) {
-    scale_colour_viridis_b(...)
+    check_scale_type(type(...), "scale_colour_binned", "colour")
   } else {
-    abort("Unknown scale type")
+    type_fallback <- getOption("ggplot2.continuous.colour", default = "gradient")
+    # don't use fallback from scale_colour_continuous() if it is
+    # a function, since that would change the type of the color
+    # scale from binned to continuous
+    if (is.function(type_fallback)) {
+      type_fallback <- "gradient"
+    }
+    type <- type %||% type_fallback
+
+    if (identical(type, "gradient")) {
+      scale_colour_steps(...)
+    } else if (identical(type, "viridis")) {
+      scale_colour_viridis_b(...)
+    } else {
+      abort("Unknown scale type")
+    }
   }
 }
 
 #' @export
 #' @rdname scale_colour_continuous
 scale_fill_binned <- function(...,
-                              type = getOption("ggplot2.binned.fill", default = getOption("ggplot2.continuous.fill", default = "gradient"))) {
+                              type = getOption("ggplot2.binned.fill")) {
   if (is.function(type)) {
-    type(...)
-  } else if (identical(type, "gradient")) {
-    scale_fill_steps(...)
-  } else if (identical(type, "viridis")) {
-    scale_fill_viridis_b(...)
+    check_scale_type(type(...), "scale_fill_binned", "fill")
   } else {
-    abort("Unknown scale type")
+    type_fallback <- getOption("ggplot2.continuous.fill", default = "gradient")
+    # don't use fallback from scale_colour_continuous() if it is
+    # a function, since that would change the type of the color
+    # scale from binned to continuous
+    if (is.function(type_fallback)) {
+      type_fallback <- "gradient"
+    }
+    type <- type %||% type_fallback
+
+    if (identical(type, "gradient")) {
+      scale_fill_steps(...)
+    } else if (identical(type, "viridis")) {
+      scale_fill_viridis_b(...)
+    } else {
+      abort("Unknown scale type")
+    }
+  }
+}
+
+
+# helper function to make sure that the provided scale is of the correct
+# type (i.e., is continuous and works with the provided aesthetic)
+check_scale_type <- function(scale, name, aesthetic) {
+  if (!is.ggproto(scale) || !inherits(scale, "Scale")) {
+    abort(glue("The `type` argument of `{name}()` must return a continuous scale for the {aesthetic} aesthetic. The provided object is not a scale function."))
+  }
+  if (!isTRUE(aesthetic %in% scale$aesthetics)) {
+    abort(glue("The `type` argument of `{name}()` must return a continuous scale for the {aesthetic} aesthetic. The provided scale works with the following aesthetics: {glue_collapse(scale$aesthetics, sep = ', ')}"))
   }
+  if (isTRUE(scale$is_discrete())) {
+    abort(glue("The `type` argument of `{name}()` must return a continuous scale for the {aesthetic} aesthetic, but the provided scale is discrete."))
```

Comment:
Shouldn't we use this for the default discrete scale as well, in which case `scale$is_discrete()` should match some input argument?

Reply:
Related issue for discrete scales: #4149

Reply:
Ok, I'll look into that issue and see if I can fix it also.

Reply:
Sorry for pilling onto what was effectively a documentation change🙂

Reply:
@thomasp85 Do you want an error or a warning when the scale doesn't contain the right aesthetic? @yutannihilation suggested a warning [here,](https://github.com/tidyverse/ggplot2/issues/4149#issuecomment-827576936) I have currently implemented an error.

Reply:
I'm slightly worried that issuing a warning and modifying the list of aesthetics can create weird cases when somebody maps two different variables to `color` and `fill`, but maybe that's a sufficiently obscure case that it doesn't matter. And there'll still be a warning that something isn't right.

Reply:
Yes, I too am worried, to be honest... The justification is that the option has a different semantics than usual `+ scale_{colour,fill}_*` in that it's used both for `fill` and for `colour` by default, in contrast to that `+ scale_{colour,fill}_*` isn't. Specifying a `scale_*` function there causes the conflict between two semantics, so apparently it needs some tweak to soften the conflict. On the other hand, the error might be a good lesson to learn `aesthetics` argument, which many users don't know how useful it is yet. I have no idea which is better.

Reply:
It looks like this is done differently for continuous and discrete scales. Continuous scales don't look at the defaults of other aesthetics. Maybe we should merge the current PR as is and defer a broader cleanup of this issue to the next major release, when we wanted to somewhat change how scales work anyways.

Reply:
I see. Sounds good to me.

Reply:
We can merge this, but I'm very tempted to also fix the discrete scales for this release. The current behaviour is a mistake I think...

At least it should be avoided when the defaults are functions and only when a palette is provided

Reply:
But to be frank that would probably be more confusing than simply not inheriting across aesthetics at all